### PR TITLE
Added bot trigger for help, location, tag

### PR DESCRIPTION
**Pull request by Js-o**
1. Able to view/update locations
2. Able to add or remove pre-existing tags (currently only 7)

### CODE REVIEW

File: server/src/routes/router.js [hunk 1]
```diff
@@ -62,7 +62,14 @@ export default (app, express) => {
   //Handling Interactive Buttons
   //////////////////////////////////////////////
   app.post('/slack/receive', buttonController.buttonDispatcher);
-  
+
+  //////////////////////////////////////////////
+  //Handling User Tags
+  //////////////////////////////////////////////
+  app.get('/slack/users/tags/:userId', userTagController.getUserTags);
```
Comment (ghost) [inline server/src/routes/router.js]: change this to /api/users/tags...

File: server/src/bot.js [hunk 2]
```diff
@@ -70,6 +71,108 @@ const askLocation = (response, convo) => {
...
+
+const respondWithTags = (bot, message) => {
+  //find all tags
+  helper.listAllTags()
+  .then(allTags => {
+    //find all user tags
+    helper.listUserTags(message)
+    .then(res => {
+      //all user tags
+      let userTag = _.map(res, item => {
```
Comment (pranaymohan) [inline server/src/bot.js]: would rename this userTags to show that its an array of tags

File: server/src/bot.js [hunk 3]
```diff
@@ -70,6 +71,108 @@ const askLocation = (response, convo) => {
...
+          actions: [button]
+        };
+        attachments.push(attachment);
+      });
+
+      let response = {
+        text: `Here are a list of your tags: `,
+        fallback: `Unable to show tags`,
+        color: `#3AA3E3`,
+        attachments: attachments
```
Comment (pranaymohan) [inline server/src/bot.js]: I think you can use shortcut to not have to repeat "a: a"

File: server/src/bot.js [hunk 4]
```diff
@@ -70,6 +71,108 @@ const askLocation = (response, convo) => {
...
+  });
+};
+
+connection.hears("location", ['direct_message'], (bot, message) => {
+  bot.reply(message, { 
+    text: `Would you like to *view* or *update* your current job search location?`,
+    attachments: [
+      {
+        text: `Choose to view or update`,
+        fallback: `You are unable to choose`,
```
Comment (pranaymohan) [inline server/src/bot.js]: Would reword this to something else -- "This option is disabled" or something softer

File: server/src/bot.js [hunk 5]
```diff
@@ -70,6 +71,108 @@ const askLocation = (response, convo) => {
...
+          }
+        ]
+      }
+    ]
+  });
+});
+
+connection.hears("help", ['direct_message'], (bot, message) => {
+  bot.reply(message, `If you would like to change your location, you can type *location*!\n` +
+    `If you would like to add filters for you search, type *tags* and toggle the button to add or ` +
```
Comment (pranaymohan) [inline server/src/bot.js]: your\* search

File: server/src/bots/helper.js [hunk 6]
```diff
@@ -42,7 +42,24 @@ const helper =  {
...
       console.log('Error: ', err);
     });
+  },
+  listUserTags: (message) => {
+
+    let userData = {
+      url: `http://localhost:8080/api/users/tags/${message.user}`,
+      method: 'GET',
+      json: true
+    }
```
Comment (pranaymohan) [inline server/src/bots/helper.js]: ;

File: server/src/bots/helper.js [hunk 7]
```diff
@@ -42,7 +42,24 @@ const helper =  {
...
+    return rp(userData);
+  },
+  listAllTags: () => {
+    let tagData = {
+      url: `http://localhost:8080/api/tags`,
+      method: `GET`,
+      json: true
+    }
```
Comment (pranaymohan) [inline server/src/bots/helper.js]: ;

File: server/src/bots/introduction.js [hunk 8]
```diff
@@ -15,13 +15,12 @@ const intro = (createdUser) => {
...
     });
   });
-
 };
 
 const askLocation = (response, convo) => {
-  convo.ask("Awesome! Where do you want to look for jobs?", (response, convo) => {
-    convo.say(`I heard that ${response.text} is a great place. ` +
-      `Well I'll be here to help you out if you need me!`);
+  convo.ask("Awesome! Which *CITY* do you want to look for jobs?", (response, convo) => {
```
Comment (pranaymohan) [inline server/src/bots/introduction.js]: CHANGE TO: Which _CITY_ would you like to search for jobs in?

File: server/src/controllers/buttonController.js [hunk 9]
```diff
@@ -10,6 +12,16 @@ const buttonDispatcher = (req, res) => {
   //dispatch to the responsible function based on the callback_id
   if (parsedPayload.callback_id === 'clickSaveJobs') {
```
Comment (pranaymohan) [inline server/src/controllers/buttonController.js:13]: might make sense to refactor to a switch-case statement

generally when you see hella else ifs, you can consider switch case

File: server/src/controllers/buttonController.js [hunk 10]
```diff
@@ -108,4 +105,124 @@ const saveJob = (req, res, data) => {
     ////////////////////////////////////////////////////////////////////
 }
 
+const locationButtons = (req, res, data) => {
+  let id = data.user.id;
+  let teamId = data.team.id;
+
+  User.find({
```
Comment (pranaymohan) [inline server/src/controllers/buttonController.js:112]: Should we continue using rp with our API instead of calling methods directly on the Model that is not associated with the controller?

File: server/src/controllers/buttonController.js [hunk 11]
```diff
@@ -108,4 +105,124 @@ const saveJob = (req, res, data) => {
...
+            attachment_type: `default`
+          }
+        ]
+      };
+      res.json(reply);
+    } else {
+      //if the button clicked was update, update the location
+      //and notify user the location was updated!
+      reply = buttonUpdater(data, "Update", 1, 'primary', 'something else');
+      reply.attachments[0].text = 'Respond to Buckleybot to update search location',
```
Comment (pranaymohan) [inline server/src/controllers/buttonController.js:139]: at some point we need to become consistent with the nomenclature of the bot. Right now we are using "Buckley", "BuckleyBot", and "Buckleybot" interchangeably

File: server/src/controllers/buttonController.js [hunk 12]
```diff
@@ -108,4 +105,124 @@ const saveJob = (req, res, data) => {
...
+      // bot asks user where they want to change location
+      store[teamId].startPrivateConversation({ user: id }, (err, convo) => {
+        convo.ask("Where do you want to change your job search location to?", (response, convo) => {
+          botHelper.updateUser(response);
+          convo.say(`Great, your location has been updated to ${response.text}!`);
+          convo.next();
+        });
+      });
+      botHelper.updateUser({ user: id, location })
```
Comment (pranaymohan) [inline server/src/controllers/buttonController.js]: ;

File: server/src/controllers/buttonController.js [hunk 13]
```diff
@@ -108,4 +105,124 @@ const saveJob = (req, res, data) => {
...
+
+const deleteUserTag = (req, res, data) => {
+  let clickedInt = `${parseInt(data.attachment_id, 10) - 1}`;
+
+  let userTagData = {
+    url: `http://localhost:8080/api/users/tags/${data.user.id}/${data.actions[0].value}`,
+    method: `DELETE`
+  }
+
+  let reply = buttonUpdater(data, 'Add Tag', 0, 'primary', 'userTag')
```
Comment (pranaymohan) [inline server/src/controllers/buttonController.js]: ;

File: server/src/controllers/buttonController.js [hunk 14]
```diff
@@ -108,4 +105,124 @@ const saveJob = (req, res, data) => {
...
+  let clickedInt = `${parseInt(data.attachment_id, 10) - 1}`;
+
+  let userId = data.user.id;
+  let tagId = data.actions[0].value;
+
+  let userTagData = {
+    url: `http://localhost:8080/api/users/tags`,
+    method: `POST`,
+    json: { userId, tagId }
+  }
```
Comment (pranaymohan) [inline server/src/controllers/buttonController.js]: ;

File: server/src/controllers/buttonController.js [hunk 15]
```diff
@@ -108,4 +105,124 @@ const saveJob = (req, res, data) => {
...
+  let reply = buttonUpdater(data, 'Delete Tag', 0, 'danger', 'userTag');
+  //update the button name to 'deleteTag'
+  //add a confirm
+  reply.attachments[clickedInt].actions[0].name = 'deleteTag';
+  reply.attachments[clickedInt].actions[0].confirm = {
+    title: `Are you sure?`,
+    text: `Confirmation to delete tag?`,
+    ok_text: `Yes, delete it!`,
+    dismiss_text: `No, don't delete!`
+  }
```
Comment (pranaymohan) [inline server/src/controllers/buttonController.js]: ;

File: server/src/controllers/buttonController.js [hunk 16]
```diff
@@ -108,4 +105,124 @@ const saveJob = (req, res, data) => {
...
+
+  rp(userTagData)
+  .then(success => console.log('Sucess adding user tag: ', success.dataValues))
```
Comment (pranaymohan) [inline server/src/controllers/buttonController.js]: Success*

File: server/src/controllers/buttonController.js [hunk 17]
```diff
@@ -108,4 +105,124 @@ const saveJob = (req, res, data) => {
...
+
+  res.json(reply);
+}
+
+const buttonUpdater = (data, buttonText, buttonInt, buttonStyle, callbackId) => {
+  let message = {
+    type: 'message',
+    text: data.original_message.text,
+    attachments: data.original_message.attachments
+  }
```
Comment (pranaymohan) [inline server/src/controllers/buttonController.js]: ;

File: server/src/controllers/buttonController.js [hunk 18]
```diff
@@ -108,4 +105,124 @@ const saveJob = (req, res, data) => {
...
+  let clickedInt = `${parseInt(data.attachment_id, 10) - 1}`;
+  console.log('clicked attachment int, ', clickedInt)
```
Comment (pranaymohan) [inline server/src/controllers/buttonController.js]: remove or ;

File: server/src/controllers/buttonController.js (hunk 10 above)
Comment (billytianlan) [reply server/src/controllers/buttonController.js:112]: Yes, this can be apart of the refactor

File: server/src/controllers/userTagController.js [hunk 19]
```diff
@@ -1 +1,38 @@
...
+    where: {
+      userId: req.params.userId,
+      tagId: req.params.tagId
+    }
+  })
+  .then(userTag => {
+    console.log('Destroyed, ', userTag.dataValues.userId, userTag.dataValues.tagId)
+    userTag.destroy();
+  })
+  .catch(err => console.log('Error destrying user tag: ', err))
```
Comment (pranaymohan) [inline server/src/controllers/userTagController.js]: ;

File: server/src/controllers/buttonController.js (hunk 10 above)
Comment (Js-o) [reply server/src/controllers/buttonController.js:112]: We have opened an issue and will be refactoring it in the coming days
